src/test/java/org/jenkinsci/remoting/protocol/ProtocolStackImplTest.java

```diff
@@ -216,6 +216,7 @@ public void pipeBasicBackToBack() throws Exception {
         west.get().send(data);
         east.get().awaitByteContent(is(expected));
         assertThat(east.get().asByteArray(), is(expected));
+        //TODO hangs with close(null)
```

Comment:
This is somewhat concerning though as `close()` calls `close(null)` but only the latter causes hanging

Reply:
I took a quick look at this. Because of polymorphism there's an important change to what the code flow looks like it should be from static analysis (going to the declaration in the IDE) and the actual flow when running / debugging. The difference is specific to the tests in how it simulates different endpoints. It looks like changing the signature [here](https://github.com/jenkinsci/remoting/blob/master/src/test/java/org/jenkinsci/remoting/protocol/IOBufferMatcherLayer.java#L50) and overriding a different method could fix it, as long as all the test usages are updated.

I'm not sure just adding the TODO comments on the usages adds more value than just leaving it as it is. Fixing it may be the best way to go, though.

Reply:
It seems like something else must be happening that causes the difference in behavior between `close()` and `close(null)`.

I assume that difference is not new behavior, but newly discovered behavior that has existed for an unknown period.  Is my assumption correct?

Reply:
Thanks for reviewing @MarkEWaite! I think your comment and mine probably crossed. The difference in behavior is due to runtime polymorphism. It's specific to the test infrastructure for creating multiple in-process connection endpoints. I'm sure it's been around for a while.

It's probably pretty easy to correct if @res0nance (or someone else) would like to spend a few minutes on it.